File: tools/serve.py
"""Static server for StudyMon.

Plain `python -m http.server` sends no cache headers, so browsers apply
heuristic caching and keep serving an old copy of the JS and CSS after the game
is updated. That looks exactly like "my change did nothing". This server tells
the browser to revalidate every time, and raises the connection backlog so a
screen that asks for hundreds of sprites at once (the Pokedex) never gets a
refused connection.

Usage:  python tools/serve.py [port]
"""
import os
import sys
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8777


class Handler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        self.send_header('Pragma', 'no-cache')
        self.send_header('Expires', '0')
        super().end_headers()

    def log_message(self, fmt, *args):
        # Only complain about failures; a Pokedex page is 1000+ successful GETs.
        if args and str(args[1]).startswith(('4', '5')):
            super().log_message(fmt, *args)


class Server(ThreadingHTTPServer):
    daemon_threads = True
    request_queue_size = 128


if __name__ == '__main__':
    os.chdir(ROOT)
    print('StudyMon serving ' + ROOT)
    print('Open http://localhost:%d   (close this window to stop)' % PORT)
    try:
        Server(('127.0.0.1', PORT), partial(Handler, directory=ROOT)).serve_forever()
    except KeyboardInterrupt:
        print('\nstopped')

File: tools/serve.js
/* Minimal static file server, used by "Play StudyMon.bat" when Python is not installed.
   Serves the StudyMon folder on 127.0.0.1 only. */
const http = require('http');
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const PORT = parseInt(process.argv[2], 10) || 8777;

const TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.jfif': 'image/jpeg',
  '.svg': 'image/svg+xml',
  '.ogg': 'audio/ogg',
  '.ico': 'image/x-icon'
};

/* ---- portrait studio save hook -------------------------------------------
   The one place this server accepts a write. It exists so tools/portrait-studio.html
   can save framing and background edits without a download dance. It writes a
   single fixed file and nothing else. Delete this block together with the
   studio when the portraits are finished. */
const EDITS = path.join(ROOT, 'tools', 'portrait-edits.json');
// Where the full body artwork lives, needed only when an edit erases background.
const ART_SOURCE = 'C:/Users/minal/Downloads/Pokemon Characters';

/* Rebuild the actual portrait files from the saved edits. Without this the
   studio only ever writes JSON and the game keeps showing the old crops. */
function applyPortraits(done) {
  const { execFile } = require('child_process');
  let edits = {};
  try { edits = JSON.parse(fs.readFileSync(EDITS, 'utf8')); } catch (e) {}
  const erasers = Object.keys(edits).filter(n => (edits[n].erase || []).length);
  const log = [];
  const reframe = () => execFile('python', ['tools/reframe-portraits.py'], { cwd: ROOT },
    (err, out, errOut) => {
      log.push(out || '', errOut || '');
      done(err ? 'failed: ' + (errOut || err.message) : log.join('').trim());
    });
  if (!erasers.length) return reframe();
  // Erase points live in the source art, so those characters are re-imported.
  execFile('python', ['tools/import-portraits.py', ART_SOURCE].concat(erasers),
    { cwd: ROOT, maxBuffer: 4e6 }, (err, out, errOut) => {
      log.push(out || '', errOut || '');
      if (err) return done('failed during import: ' + (errOut || err.message));
      reframe();
    });
}

http.createServer((req, res) => {
  let rel = decodeURIComponent(req.url.split('?')[0]);
  if (req.method === 'POST' && rel === '/portrait-apply') {
    applyPortraits(msg => {
      res.writeHead(msg.indexOf('failed') === 0 ? 500 : 200,
        { 'Content-Type': 'text/plain; charset=utf-8' }).end(msg);
    });
    return;
  }
  if (req.method === 'POST' && rel === '/portrait-edits') {
    let body = '';
    req.on('data', c => { body += c; if (body.length > 4e6) req.destroy(); });
    req.on('end', () => {
      try {
        JSON.parse(body);                       // refuse anything that is not JSON
        fs.writeFileSync(EDITS, body);
        res.writeHead(200, { 'Content-Type': 'text/plain' }).end('saved');
      } catch (e) {
        res.writeHead(400, { 'Content-Type': 'text/plain' }).end('bad json');
      }
    });
    return;
  }
  if (rel === '/' || rel === '') rel = '/index.html';
  const file = path.join(ROOT, path.normalize(rel));
  // never serve anything outside the game folder
  if (!file.startsWith(ROOT)) { res.writeHead(403).end('forbidden'); return; }
  fs.readFile(file, (err, buf) => {
    if (err) { res.writeHead(404, { 'Content-Type': 'text/plain' }).end('not found'); return; }
    res.writeHead(200, {
      'Content-Type': TYPES[path.extname(file).toLowerCase()] || 'application/octet-stream',
      // Revalidate every time, so an updated game file is never served stale.
      'Cache-Control': 'no-cache, no-store, must-revalidate'
    });
    res.end(buf);
  });
}).listen(PORT, '127.0.0.1', () => {
  console.log('StudyMon serving ' + ROOT);
  console.log('Open http://localhost:' + PORT + '  (Ctrl+C to stop)');
});
